### PR TITLE
Anonymous ab File Backed

### PR DESCRIPTION
**Pull request by Guy-Markman**
@alonbl

### CODE REVIEW

**Comment by Guy-Markman** [comment]
@alonbl
Comments? May I merge?

File: EX06/file_backed_1.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+def main():
+    fd = os.open(
+        constants.BACKUP_FILE,
+        os.O_CREAT | os.O_RDWR,
+        stat.S_IREAD | stat.S_IWRITE | stat.S_IRGRP | stat.S_IROTH
+    )
+
+    os.lseek(fd, constants.FILE_SIZE, os.SEEK_SET)
+    os.write(fd, "\0")
```
Comment (alonbl) [inline EX06/file_backed_1.py]: please make sure that the file is at the right size

File: EX06/file_backed_1.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    try:
+        with contextlib.closing(mmap.mmap(fd, 1024)) as mm:
```
Comment (alonbl) [inline EX06/file_backed_1.py]: constant?

File: EX06/file_backed_1.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            counter = 0
+            while True:
+                mm[
+                    constants.START_LOCATION:
+                    constants.END_LOCATION
+                ] = "%4x" % counter
```
Comment (alonbl) [inline EX06/file_backed_1.py]: you have nice constants for start and end, but use hardcoded number for the format... if you change one you must change the other... not best solution.

File: EX06/file_backed_2.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def proc_child(file):
+    while True:
+        util.write_to_target(
+            1,
+            "%s\n" % file[constants.START_LOCATION:constants.END_LOCATION]
+        )
+        time.sleep(2)
+        sys.close()
```
Comment (alonbl) [inline EX06/file_backed_2.py]: what is this?

File: EX06/file_backed_2.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def proc_parent(file):
+    counter = 0
+    while True:
+        hex_counter = hex(counter)
```
Comment (alonbl) [inline EX06/file_backed_2.py]: never use hex

File: EX06/file_backed_2.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            len(hex_counter)] = hex_counter
+        counter += 1
+        time.sleep(1)
+        sys.close()
+
+
+def main():
+    fd = os.open(constants.BACKUP_FILE, os.O_RDWR)
+    try:
+        mm = mmap.mmap(fd, 0)
```
Comment (alonbl) [inline EX06/file_backed_2.py]: incorrect use of contextlib.closing

File: EX06/file_backed_1.py [hunk 7]
```diff
@@ -14,16 +14,23 @@ def main():
         stat.S_IREAD | stat.S_IWRITE | stat.S_IRGRP | stat.S_IROTH
     )
 
-    os.lseek(fd, constants.FILE_SIZE, os.SEEK_SET)
+    os.lseek(fd, constants.FILE_SIZE-1, os.SEEK_SET)
+    before_changes = None
+    if os.fsta(fd).st_size >= constants.FILE_SIZE: # TODO: Change it so it'll work in both cases
```
Comment (alonbl) [inline EX06/file_backed_1.py]: why is it needed?

File: EX06/file_backed_1.py [hunk 8]
```diff
@@ -14,16 +14,23 @@ def main():
...
+        os.lseek(fd, constants.START_LOCATION, os.SEEK_SET)
+        before_changes = os.read
```
Comment (alonbl) [inline EX06/file_backed_1.py]: not sure why this before_changes is required, please think simple!

Comment (Guy-Markman) [inline EX06/file_backed_1.py]: I was overthinking. I wanted to return the file to the original state.

Comment (Guy-Markman) [inline EX06/file_backed_1.py]: I tried that it'll work also on a file which was already in the system when we started.

File: EX06/file_backed_1.py [hunk 9]
```diff
@@ -6,20 +6,18 @@
...
         stat.S_IREAD | stat.S_IWRITE | stat.S_IRGRP | stat.S_IROTH
     )
-
-    os.lseek(fd, constants.FILE_SIZE-1, os.SEEK_SET)
-    before_changes = None
-    if os.fsta(fd).st_size >= constants.FILE_SIZE: # TODO: Change it so it'll work in both cases
-        os.lseek(fd, constants.START_LOCATION, os.SEEK_SET)
-        before_changes = os.read
-    os.write(fd, "\0")
+    while os.stat(fd).st_size < constants.FILE_SIZE:
```
Comment (alonbl) [inline EX06/file_backed_1.py]: you should probably start with a fresh file every time

File: EX06/constants.py [hunk 10]
```diff
@@ -1,5 +1,5 @@
-BACKUP_FILE = "backup"
-END_LOCATION = 14
-FILE_SIZE = 1024
-ORIGINAL_FILE = "origin"
-START_LOCATION = 10
+BACKUP_FILE = "backup"
+END_LOCATION = 14
+FILE_SIZE = 1024
+ORIGINAL_FILE = "origin"
```
Comment (alonbl) [inline EX06/constants.py]: ?

File: EX06/constants.py [hunk 11]
```diff
@@ -1,5 +1,5 @@
-BACKUP_FILE = "backup"
-END_LOCATION = 14
-FILE_SIZE = 1024
-ORIGINAL_FILE = "origin"
-START_LOCATION = 10
+BACKUP_FILE = "backup"
```
Comment (alonbl) [inline EX06/constants.py]: ?

**Comment by Guy-Markman** [comment]
I tried to merge two commits as good as I can,
Didn't end well

File: EX06/file_backed_1.py [hunk 12]
```diff
@@ -15,7 +15,11 @@ def main():
         os.O_CREAT | os.O_RDWR,
         stat.S_IREAD | stat.S_IWRITE | stat.S_IRGRP | stat.S_IROTH
     )
+<<<<<<< HEAD
+    while os.fstat(fd).st_size < constants.FILE_SIZE:
+=======
     while os.stat(fd).st_size < constants.FILE_SIZE:
+>>>>>>> 232deea8a33dd1896f0463c5638bb8802d543bbc
```
Comment (Guy-Markman) [inline EX06/file_backed_1.py]: I tried to merge.
I just uploaded the fix

**Pull request merged**